Add support for `no_new_privs` via AllowPrivilegeEscalation

jessfraz opened the pull request:
**What this PR does / why we need it**:
Implements kubernetes/community#639
Fixes #38417

Adds `AllowPrivilegeEscalation` and `DefaultAllowPrivilegeEscalation` to `PodSecurityPolicy`.
Adds `AllowPrivilegeEscalation` to container `SecurityContext`.

Adds the proposed behavior to `kuberuntime`, `dockershim`, and `rkt`. Adds a bunch of unit tests to ensure the desired default behavior and that when `DefaultAllowPrivilegeEscalation` is explicitly set.

Tests pass locally with docker and rkt runtimes. There are also a few integration tests with a `setuid` binary for sanity.

**Release note**:

```release-note
Adds AllowPrivilegeEscalation to control whether a process can gain more privileges than it's parent process
```

jessfraz committed:
allowPrivilegeEscalation: add integration test with setuid binary

Signed-off-by: Jess Frazelle <acidburn@google.com>

dims commented:
@jessfraz : sorry for a dumb question. Is this feature meant to only be active/available when the kubelet `--allow-privileged` is true?

jessfraz commented:
@dims there is a chart in the proposal that might help, `no_new_privs` will be added when allowPrivilegeEscalation is false, by default this will be when a container is not privileged, adding CAP{_SYS_ADMIN, or when the uid!=0

dims commented:
@jessfraz got it. thanks!

jessfraz commented:
Tests pass locally with `docker` and `rkt` runtimes.

jessfraz commented on test/e2e_node/image_list.go:
```diff
@@ -53,6 +53,7 @@ var NodeImageWhiteList = sets.NewString(
 	"gcr.io/google_containers/nginx-slim:0.7",
 	"gcr.io/google_containers/serve_hostname:v1.4",
 	"gcr.io/google_containers/netexec:1.7",
+	"r.j3ss.co/no_new_privs:latest",
```
I will switch this to `google_containers` before merge don't worry

jessfraz replied:
fixed the image to be on google container registry

pweil- commented:
@php-coder

jessfraz committed:
test/images: add no_new_privs test container

Using the image:

```
$ docker run --rm -it --user 1000 gcr.io/google_containers/nonewprivs:1.0
Effective uid: 0

$ docker run --rm -it --user 1000 --security-opt no-new-privileges gcr.io/google_containers/nonewprivs:1.0
Effective uid: 1000
```

Signed-off-by: Jess Frazelle <acidburn@google.com>

pweil- commented on pkg/securitycontext/util.go:
```diff
@@ -134,6 +134,11 @@ func DetermineEffectiveSecurityContext(pod *v1.Pod, container *v1.Container) *v1
 		*effectiveSc.ReadOnlyRootFilesystem = *containerSc.ReadOnlyRootFilesystem
 	}
 
+	if containerSc.AllowPrivilegeEscalation != nil {
```
:+1: on finding this, the `DetermineEffectiveSecurityContext` will cause subtle issues in PSP if not handled (and I've forgotten it a number of times)

jessfraz replied:
I felt relief after I found it too hahahahaaha

jessfraz commented:
The test failure is not the test I added so whoop lol *cries*

jessfraz commented:
@k8s-bot pull-kubernetes-node-e2e test this

jessfraz committed:
allowPrivilegeEscalation: apply to correct docker api versions

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz commented:
Ok yay they all passed

jessfraz commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

lucab commented on test/images/nonewprivs/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM alpine:latest
+
+COPY nnp /usr/local/bin/nnp
+RUN chmod +s /usr/local/bin/nnp
```
I see you want this compiled as a static ELF in the Makefile. If that's the case, can we avoid an alpine dependency and a spurious layer and just use `FROM: scratch` plus `sudo chmod` on the host side? (Other test containers in here seems to do so)

jessfraz replied:
Can update

jessfraz replied:
this wont work if its not done like this

lucab replied:
No worry, this was just a minor suggestion. Is the sticky bit dropped during dockerbuild COPY?

jessfraz replied:
yeah i am assuming it's that didn't have time to debug

lucab commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -933,6 +933,15 @@ type PodSecurityPolicySpec struct {
 	AllowedHostPaths []string `json:"allowedHostPaths,omitempty" protobuf:"bytes,15,opt,name=allowedHostPaths"`
+	// DefaultAllowPrivilegeEscalation controls the default setting for whether a
+	// process can gain more privileges than it's parent process.
+	// // +optional
```
I hope this is the source location, because those are propagated to all auto-generated snippets:
 * s/it's parent/its parent/
 * Spurious double slash before the optional annotation

The first typo is also in PR comment / release note, but I can't comment directly there :)

php-coder replied:
@jessfraz I see that s/it's parent/its parent/ wasn't fixed. Do you want me to prepare PR for that?

php-coder replied:
I've submitted https://github.com/kubernetes/kubernetes/pull/51536

lucab commented on test/e2e_node/security_context_test.go:
```diff
@@ -274,4 +276,84 @@ var _ = framework.KubeDescribe("Security Context", func() {
 
+	Context("when creating containers with AllowPrivilegeEscalation", func() {
+		// parse the docker version
+		out, err := exec.Command("docker", "-v").CombinedOutput()
```
There is a `ContainerRuntime` in `framework/TestContext` to make this conditional and effective only for docker test environments.

jessfraz replied:
Thanks I looked all over

timstclair replied:
Better yet, copy [SkipIfContainerRuntimeIs](https://github.com/kubernetes/kubernetes/blob/afc4506c33063fe64c840f2696af42bfb996436c/test/e2e/framework/util.go#L348) to make `SkipUnlessContainerRuntimeIs`.

This should work for rocket too, right?

jessfraz replied:
this test works with rkt tho, wouldnt that make it skip it if it wasn't docker?

jessfraz commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -933,6 +933,15 @@ type PodSecurityPolicySpec struct {
 	AllowedHostPaths []string `json:"allowedHostPaths,omitempty" protobuf:"bytes,15,opt,name=allowedHostPaths"`
+	// DefaultAllowPrivilegeEscalation controls the default setting for whether a
+	// process can gain more privileges than it's parent process.
+	// // +optional
```
Gotcha thanks

timstclair commented on pkg/security/podsecuritypolicy/provider.go:
```diff
@@ -310,6 +316,10 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
 		}
 	}
 
+	if !s.psp.Spec.AllowPrivilegeEscalation && sc.AllowPrivilegeEscalation != nil && *sc.AllowPrivilegeEscalation {
```
I think this always happens after defaulting, which means we should require that `AllowPrivilegeEscalation != nil`. Otherwise if the admin forgets to set a default and the user leaves this empty, that's equivalent to allowing privilige escalation in certain scenarios (IIUC). Alternatively, we could default to false if it's prohibited by the psp and no default is set.

jessfraz replied:
we don't have a default we set we just have a behavior though when it is nil (proposal: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/no-new-privs.md#pod-security-policy-changes), if I understand your comment correctly, so it is allowed to be nil here I think

timstclair replied:
Let me see if I can describe the scenario better:

1. Admin sets `psp.AllowPrivilegeEscalation = false`, but leaves `psp.DefaultAllowPrivilegeEscalation = nil`.
2. User creates a pod with a non-root user, and leaves `pod.Spec.SecurityContext.AllowPrivilegeEscalation = nil`.
3. Validation check here lets it through, because it is nil.
4. Kubelet defaults it to allowed

As an admin, I would expect that the value I set in (1) would enforce that `AllowPrivilegeEscalation` is always false (ignoring the privileged pod case), and the result in (4) would surprise me.

jessfraz replied:
ah I see

jessfraz replied:
so `psp.AllowPrivilegeEscalation` is just "am i allowed to set securityContext.AllowPrivilegeEscalation" so I think it might be a naming problem

jessfraz replied:
the naming convention was taken from psp.Privileged which I also agree is confusing

timstclair commented on pkg/securitycontext/util.go:
```diff
@@ -232,3 +237,51 @@ func internalSecurityContextFromPodSecurityContext(pod *api.Pod) *api.SecurityCo
+// 1) the container is not privileged
+// 2) CAP_SYS_ADMIN is not being added
+// 3) if podSecurityPolicy.DefaultAllowPrivilegeEscalation is:
+//		- nil, then return true if uid != 0
```
This will still break SUID binaries if the uid is set in the Dockerfile rather than through k8s, right?

jessfraz replied:
yeah, if the default is to use whatever the dockerfile is if not set thru the k8s communication with the docker API, then yes, I am not familiar with the behavior tho

timstclair replied:
Yes, if the user isn't set through k8s, the value from the dockerfile is used. I know what you have here is consistent with the proposal, but I'm worried about breaking users with this. It seems like we're already turning it off in most of the cases where it would actually be useful as a security primitive, right? In that case, can we just leave it off in the default case? (i.e. if `pod.Spec..AllowPrivilegeEscalation == nil && psp.DefaultAllowPrivilegEscalation == nil` then the container is run without `no_new_privs`?)

jessfraz replied:
yeah this makes sense

timstclair commented on test/images/nonewprivs/Makefile:
```diff
@@ -0,0 +1,33 @@
+.PHONY: all image push clean
+
+TAG = 1.0
+PREFIX = gcr.io/google_containers
```
nit: s/google_containers/google-containers/

jessfraz replied:
i think they are all google_containers? should they all change?

timstclair replied:
Eh, you can leave it. We're gradually transitioning everything to `google-containers`, but we should probably just send a separate PR to fix all of these.

timstclair commented on test/e2e_node/security_context_test.go:
```diff
@@ -274,4 +276,84 @@ var _ = framework.KubeDescribe("Security Context", func() {
+		}
+
+		It("should allow privilege escalation when not explicitly set and uid != 0", func() {
+			podName := "alpine-nnp-" + string(uuid.NewUUID())
```
nit: Just use the `GenerateName` on the pod spec.

php-coder replied:
Unfortunately, it's impossible because we need to know a pod name to be able to wait for it and inspect its logs later.

jessfraz committed:
allowPrivilegeEscalation: modify api types & add functionality

Signed-off-by: Jess Frazelle <acidburn@google.com>

timstclair commented on pkg/securitycontext/util.go:
```diff
@@ -232,3 +237,51 @@ func internalSecurityContextFromPodSecurityContext(pod *api.Pod) *api.SecurityCo
+	}
+
+	// handle the case where defaultAllowPrivilegeEscalation is false or the user explicitly set allowPrivilegeEscalation to false
+	if sc.AllowPrivilegeEscalation != nil && !*sc.AllowPrivilegeEscalation {
```
nit: simplify to:
```go
if sc.AllowPrivilegeEscalation != nil {
  return !*sc.AllowPrivilegeEscalation
}
```

jessfraz committed:
allowPrivilegeEscalation: update docs

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz committed:
allowPrivilegeEscalation: update code generation

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz commented:
/test pull-kubernetes-federation-e2e-gce

jessfraz commented:
all green and rebased

tallclair commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -113,6 +113,14 @@ func modifyHostConfig(sc *runtimeapi.LinuxContainerSecurityContext, hostConfig *
+	if sc.NoNewPrivs {
+		// "no-new-privileges" was added in Docker 1.11
+		// which is API version 1.23
+		if apiVersion.GTE(semver.Version{Major: 1, Minor: 23}) {
```
Should we block the pod from running if `no-new-privileges` can't be enforced? That's what we do for apparmor.  (https://github.com/kubernetes/kubernetes/blob/3fb2d8248d7f367758292ac0964cf0d33a8f4678/pkg/kubelet/kubelet.go#L812-L813)

jessfraz replied:
oh interesting... that would be cool

jessfraz replied:
As long as it doesn't break people I think it should be ok to do

jessfraz replied:
So I am not sure exactly how we would do this other than checking the kernel version is >= 3.5

tallclair replied:
It doesn't need to be perfect... You could start with `runtime == rkt || (runtime == docker && apiversion >= 1.23)`

jessfraz replied:
updated

jessfraz replied:
should I get rid of this check then?

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4613,6 +4613,15 @@ type SecurityContext struct {
+	// By default, AllowPrivilegeEscalation is false unless the container is
+	// 1) run as Privileged
+	// 2) has CAP_SYS_ADMIN
+	// 3) UID of the container is not 0
```
I believe this is no longer the case?

jessfraz replied:
wah good catch thought i got them alll

tallclair reviewed (commented):
Just 1 nit & 1 question, then LGTM.

cblecker commented:
/lint

spxtr commented:
/lint

spxtr commented:
/lint

spxtr commented:
/lint

jessfraz commented:
It works in rkt :)

On Jul 20, 2017 17:21, "Tim Allclair (St. Clair)" <notifications@github.com>
wrote:

> *@tallclair* commented on this pull request.
> ------------------------------
>
> In pkg/kubelet/dockershim/security_context.go
> <https://github.com/kubernetes/kubernetes/pull/47019#discussion_r128633883>
> :
>
> > @@ -113,6 +113,14 @@ func modifyHostConfig(sc *runtimeapi.LinuxContainerSecurityContext, hostConfig *
>  	}
>  	hostConfig.SecurityOpt = append(hostConfig.SecurityOpt, apparmorSecurityOpts...)
>
> +	if sc.NoNewPrivs {
> +		// "no-new-privileges" was added in Docker 1.11
> +		// which is API version 1.23
> +		if apiVersion.GTE(semver.Version{Major: 1, Minor: 23}) {
>
> It doesn't need to be perfect... You could start with runtime == rkt ||
> (runtime == docker && apiversion >= 1.23)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47019#discussion_r128633883>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbAjv_wnokfOfBMo3ECnvmj8nPMxrks5sP8S9gaJpZM4Nw1Aj>
> .
>

jessfraz commented:
/retest

jessfraz commented:
/test pull-kubernetes-e2e-gce-etcd3

jessfraz commented:
/test pull-kubernetes-e2e-gce-etcd3

tallclair commented on pkg/kubelet/lifecycle/handlers.go:
```diff
@@ -165,3 +165,75 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+	}
+
+	// Make sure it is either docker or rkt runtimes.
+	if a.Runtime.Type() != kubetypes.DockerContainerRuntime && a.Runtime.Type() != kubetypes.RktContainerRuntime {
```
@yujuhong - It would be nice to incorporate this type of logic into the CRI. What if there was something like "GetSupportedFeatures", which would return a list of optional features (known constants) that the CRI runtime supported?

tallclair commented on pkg/kubelet/lifecycle/handlers.go:
```diff
@@ -165,3 +165,75 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+		}
+	}
+
+	// Make sure docker api version is valid.
```
move this into the if block below, after you check whether the runtime is docker.

spxtr commented:
/lint

spxtr commented:
Lint comments on generated code aren't too useful :disappointed:

jessfraz commented:
/retest

jessfraz commented:
this is all updated and green :)

tallclair commented:
/lgtm

jessfraz commented:
/retest

On Jul 28, 2017 20:29, "k8s-ci-robot" <notifications@github.com> wrote:

> @jessfraz <https://github.com/jessfraz>: The following test *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-kops-aws a5e4c6f
> <https://github.com/kubernetes/kubernetes/commit/a5e4c6f6ed0d441d8fd17a18416556f7621eeb56>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/47019/pull-kubernetes-e2e-kops-aws/37905/> /test
> pull-kubernetes-e2e-kops-aws
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/47019>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/jessfraz>. Please
> help us cut down on flakes by linking to
> <https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://github.com/kubernetes/test-infra/blob/master/commands.md>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47019#issuecomment-318790100>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbAH6PtUVf8iRfE-Y0MPavb0NcJfuks5sSnz8gaJpZM4Nw1Aj>
> .
>

jessfraz commented:
ping @smarterclayton for approval 😇

smarterclayton commented:
API approved as per the proposal and discussion on the thread.  

/approve

jessfraz commented:
/test pull-kubernetes-bazel

mkumatag commented on test/images/nonewprivs/nnp.c:
```diff
@@ -0,0 +1,22 @@
+// Copyright 2017 The Kubernetes Authors.
```
Is there any specif reason my we are using C code? I see even os library in go has `Geteuid()` which can be used.

jessfraz replied:
Does it matter?

mkumatag replied:
Yes. For multi arch test images we are planning to have all images under `kubernetes/tests/images` directory for supported architectures like `amd64, arm, arm64, ppc64le`. So its always very easy to cross build the go code over a C code.

grodrigues3 commented:
Which sig does this fall under?

jessfraz commented:
sig-auth

On Aug 16, 2017 13:48, "grodrigues3" <notifications@github.com> wrote:

> Which sig does this fall under?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47019#issuecomment-322847905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbK92W_zltYwO7qZXMwUV2LuPWlcsks5sYyuFgaJpZM4Nw1Aj>
> .
>

jessfraz commented:
Sure thanks

On Aug 29, 2017 08:56, "Vyacheslav Semushin" <notifications@github.com>
wrote:

> *@php-coder* commented on this pull request.
> ------------------------------
>
> In pkg/apis/extensions/v1beta1/types.go
> <https://github.com/kubernetes/kubernetes/pull/47019#discussion_r135780591>
> :
>
> > @@ -933,6 +933,15 @@ type PodSecurityPolicySpec struct {
>  	// host paths may be used.
>  	// +optional
>  	AllowedHostPaths []string `json:"allowedHostPaths,omitempty" protobuf:"bytes,15,opt,name=allowedHostPaths"`
> +	// DefaultAllowPrivilegeEscalation controls the default setting for whether a
> +	// process can gain more privileges than it's parent process.
> +	// // +optional
>
> @jessfraz <https://github.com/jessfraz> I see that s/it's parent/its
> parent/ wasn't fixed. Do you want me to prepare PR for that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47019#discussion_r135780591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbBeUnZqwIVlGeDLR3xA0vePnRL7_ks5sdAqSgaJpZM4Nw1Aj>
> .
>

php-coder commented on pkg/security/podsecuritypolicy/provider_test.go:
```diff
@@ -1033,3 +1034,111 @@ func TestValidateAllowedVolumes(t *testing.T) {
+func TestValidateAllowPrivilegeEscalation(t *testing.T) {
+	pod := defaultPod()
+	pe := true
+	pod.Spec.Containers[0].SecurityContext.AllowPrivilegeEscalation = &pe
```
I see that tests don't cover that case of using InitContainers? Does this validation work for them? I assume that it should but I don't see any mentions about them in the code.

php-coder commented on pkg/kubelet/lifecycle/handlers.go:
```diff
@@ -165,3 +165,74 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+	return PodAdmitResult{Admit: true}
+}
+
+func noNewPrivsRequired(pod *v1.Pod) bool {
```
@jessfraz Should we handle `InitContainers` here too? Looks like, yes, we should.

jessfraz replied:
Are init containers handled differently that should probably be documented somewhere so it's not overlooked unless I missed it

php-coder replied:
As far I understand, the most of the code from this PR is handling initContainers correctly because it works only with securityContext. The existing infrastructure is aware about containers/initContainers and executes validation on their SecurityContexts.

But when we're manually inspecting PodSpec we obligated to take care about initContainers. In this particular case, I see that `noNewPrivsRequired()` will return `false` for a pod that has initContainers with allowPrivielegeEscalation=false. I'm not sure what will be the consequence of this, perhaps `allowPrivielegeEscalation=false` won't work for a pod?

php-coder commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4613,6 +4613,14 @@ type SecurityContext struct {
+	// the no_new_privs flag will be set on the container process.
+	// AllowPrivilegeEscalation is true always when the container is:
+	// 1) run as Privileged
+	// 2) has CAP_SYS_ADMIN
```
@jessfraz Are we missing the case "when a container is not run as root" here?

jessfraz replied:
We got rid of that functionality

php-coder replied:
And the proposal is outdated now? https://github.com/kubernetes/community/blob/f149433600ff1795521e7d9f2ae8ff3162394c61/contributors/design-proposals/no-new-privs.md#changes-of-securitycontext-objects

php-coder replied:
Also some tests can be deleted then: https://github.com/kubernetes/kubernetes/blob/28f6b3fcc037c3a21048160d062f6d544081bd71/pkg/securitycontext/util_test.go#L203-L212
